doc/reference/trd104-syscalls.md

```diff
@@ -553,21 +556,104 @@ an error code of `INVALID`.
+
+```rust
+if mlen >= arg1 && arg1 > 0 {
+    app.len = arg1;
+    app.index = 0;
+    self.busy.set(true);
+    self.do_next_read_write(app);
+    CommandReturn::success()
+} 
+```
+
+Checking that the length fits within the allowed buffer when the
+command is issued is insufficient, as it could be that the buffer
+changes during the underlying hardware I/O operation.  If the buffer
+is replaced with one that is much smaller, the length passed in the
+command may now be too large. The `index` variable keeps track of
+where in the buffer the next write should occur: the capsule breaks up
+long writes into multiple, smaller writes to bound the size of its
+static kernel buffer. If capsule code blindly copies the number of
+bytes specified in the command, without re-checking buffer length,
+then it can cause the kernel to panic for an out-of-bounds error. 
+
+Therefore, in the `read_write_done` callback, the capsule checks the
+length of the buffer that userspace wants to read data into. The third
+line checks that the end of the just completed operation isn't past
+the end of the current userspace buffer (which could happen if the
+userspace buffer became shorter).
+
+```rust
+let end = index;
+let start = index - length;
+let end = cmp::min(end, dest.len());
+let start = cmp::min(start, end);
+
+let real_len = cmp::min(end - start, src.len());
+let dest_area = &mut dest[start..end];
+
+for (i, c) in src[0..real_len].iter().enumerate() {
+    dest_area[i] = *c;
+}
+```
+
+For similar reasons, a capsule should not cache computations on values
+from an allowed buffer, as if the buffer changes those computations
+may no longer be correct (e.g., computing a length based on fields in
+the buffer).
+
 4.5 Read-Only Allow (Class ID: 4)
 ---------------------------------
 
-The Read-Only Allow class is very similar to the Read-Write Allow class.
-It differs in tow ways: the buffer it passes to the kernel is read-only,
-and the process retains read access to the buffer. The kernel cannot
-write to the buffer. The semantics and calling conventions of
-Read-Only Allow are otherwise identical to Read-Write Allow.
+The Read-Only Allow class is very similar to the Read-Write Allow
+class.  It differs in two ways: the buffer it passes to the kernel is
+read-only, and the process MAY freely read the buffer. The kernel
+also MUST NOT write to a buffer shared with a Read-Only Allow. The
+semantics and calling conventions of Read-Only Allow are otherwise
+identical to Read-Write Allow: a userspace API MUST NOT depend on writing 
+to a shared buffer and the kernel MUST NOT assume the buffer does 
+not change.
+
+This restriction on writing to buffers is to limit the complexity of
+code review in the kernel. If a userspace library relies on writes to
+shared buffers, then kernel code correspondingly relies on them. This
+sort of concurrent access can have unforeseen edge cases which cause
+the kernel to panic, e.g., because values changed between method
+calls.
```

Comment:
A second consideration here is also that a userspace API which requires this interface would complicate the libtock-rs API considerably -- that might be worth mentioning here as well.